utils/testtools/crypto.go

```diff
@@ -135,72 +79,138 @@ func CreateCSR(key crypto.PrivateKey) (csr []byte, err error) {
+		return nil, nil, err
 	}
 
-	if err = ioutil.WriteFile(csrConfFile, []byte(csrConf), filePerm); err != nil {
-		return nil, aoserrors.Wrap(err)
+	return cert, key, nil
+}
+
+// Verify verify chain of certificates.
+func Verify(leafCert *x509.Certificate, roots, intermediates []*x509.Certificate) error {
+	rootPool := x509.NewCertPool()
+	intermediatePool := x509.NewCertPool()
+
+	for _, c := range roots {
+		rootPool.AddCert(c)
 	}
 
-	var out []byte
+	for _, c := range intermediates {
+		intermediatePool.AddCert(c)
+	}
 
-	if out, err = exec.Command("openssl", "req", "-inform", "PEM", "-in",
-		csrFile, "-out", csrFile+".pem").CombinedOutput(); err != nil {
-		return nil, aoserrors.Errorf("message: %s, %s", string(out), err)
+	opts := x509.VerifyOptions{
+		Roots:         rootPool,
+		Intermediates: intermediatePool,
 	}
 
-	if out, err = exec.Command("openssl", "x509", "-req", "-in", csrFile+".pem",
-		"-CA", caCertFile, "-CAkey", caKeyFile, "-extfile", csrConfFile, "-extensions", "ext",
-		"-outform", "PEM", "-out", unitCertFile, "-CAcreateserial", "-days", "3650").CombinedOutput(); err != nil {
-		return nil, aoserrors.Errorf("message: %s, %s", string(out), err)
+	if _, err := leafCert.Verify(opts); err != nil {
+		return aoserrors.Wrap(err)
 	}
 
-	certData, err := ioutil.ReadFile(unitCertFile)
+	return nil
+}
+
+// GenerateDefaultCARootCertAndKey generate default CA root certificate and key.
+func GenerateDefaultCARootCertAndKey() (cert *x509.Certificate, key *rsa.PrivateKey, err error) {
+	serialNumber, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 128))
 	if err != nil {
-		return nil, aoserrors.Wrap(err)
+		return nil, nil, aoserrors.Wrap(err)
 	}
 
-	caData, err := ioutil.ReadFile(caCertFile)
+	rootTemplate := DefaultCertificateTemplate
+	rootTemplate.SerialNumber = serialNumber
+	rootTemplate.IsCA = true
+	rootTemplate.Subject.OrganizationalUnit = []string{"Novus Ordo Seclorum"}
+	rootTemplate.Subject.CommonName = "Fusion Root CA"
+
+	cert, key, err = GenerateKeyAndCertificate(&rootTemplate, &rootTemplate, nil)
 	if err != nil {
-		return nil, aoserrors.Wrap(err)
+		return nil, nil, aoserrors.Wrap(err)
+	}
+
+	return cert, key, nil
+}
+
+// GenerateCACertPEMAndKey generate CA certificate and key.
+func GenerateCACertAndKey(
+	parent *x509.Certificate, privateKey *rsa.PrivateKey, subject pkix.Name,
+) (*x509.Certificate, *rsa.PrivateKey, error) {
+	serialNumber, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 128))
+	if err != nil {
+		return nil, nil, aoserrors.Wrap(err)
+	}
+
+	template := DefaultCertificateTemplate
+	template.SerialNumber = serialNumber
+	template.Subject = subject
+	template.IsCA = true
+
+	cert, key, err := GenerateKeyAndCertificate(&template, parent, privateKey)
+	if err != nil {
+		return nil, nil, aoserrors.Wrap(err)
+	}
+
+	return cert, key, nil
+}
+
+// GenerateCertAndKey generate certificate and key.
+func GenerateCertAndKey(
```

Comment:
ditto

And please rename it, we have `GenerateKeyAndCertificate` already.

Reply:
Done